server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java

```diff
@@ -28,19 +29,45 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.function.Function;
 import java.util.stream.Stream;
 
-public final class MappingLookup {
+public class MappingLookup {
+    /**
+     * Key for the lookup to be used in caches.
+     */
+    public class CacheKey {
+        private CacheKey() {}
+    }
+
+    /**
+     * A lookup representing an empty mapping.
+     */
+    public static final MappingLookup EMPTY = new MappingLookup(
+        "_doc",
+        List.of(),
+        List.of(),
+        List.of(),
+        List.of(),
+        0,
+        souceToParse -> null,
+        false
+    );
+
+    private final CacheKey cacheKey = new CacheKey();
+
     /** Full field name to mapper */
     private final Map<String, Mapper> fieldMappers;
     private final Map<String, ObjectMapper> objectMappers;
     private final boolean hasNested;
     private final FieldTypeLookup fieldTypeLookup;
     private final int metadataFieldCount;
     private final Map<String, NamedAnalyzer> indexAnalyzers = new HashMap<>();
+    private final Function<SourceToParse, ParsedDocument> documentParser;
```

Comment:
@javanna and I talked earlier today and we thought that this function hid a volatile read. I *think* it doesn't. It is genuinely weird to be able to parse documents at search time but that is how the percolator does its thing.

Reply:
I also don't think it hides a volatile read, but I do find the logic hard to follow: this function uses `DocumentMapper`, which is a holder object with more state. For this change I think it'd be cleanest to remove this function parameter and use `MapperService` directly to perform parsing: `mapperService.documentMapper().parse(...)`. Then we can figure out how to remove this as we work through the other follow-ups.

Reply:
I am also under the impression that we could do without a function, which ends up hiding the dependency on DocumentMapper, though I like that the dependency is then on a single method rather than on all of DocumentMapper or MapperService which end up exposing much more stuff that is not needed and I would not want to end up relying on.

I have been trying to simplify this by removing the need for the whole document mapper. I was not incredibly sure but I gave it a try with #66780 and I think we may be able to use DocumentParser directly here, which I think may make us all happier.

For the sake of this PR, I don't have a strong opinion but I am slightly in favour of leaving the function knowing that we have plans to replace it. If you feel strongly about `mapperService.documentMapper().parse(...)` I won't object ;)

Reply:
I don't have strong feelings, given we are tracking plans to replace it! I had missed that when I tried to summarize the follow-ups.